File: 传统的设计模式/抽象工厂.cpp
#include<iostream>
using namespace std;

class Fruit
{
public:
	virtual void sayName() = 0;
};



class NorthBanana :public Fruit
{
public:
	void sayName()
	{
		cout <<"I am NorthBanana" << endl;
	}
};

class NorthApple :public Fruit
{
public:
	void sayName()
	{
		cout << "I am NorthApple" << endl;
	}
};

class SouthApple :public Fruit
{
public:
	void sayName()
	{
		cout << "I am SouthApple" << endl;
	}
};

class SouthBanana :public Fruit
{
public:
	void sayName()
	{
		cout << "I am SouthBanana" << endl;
	}
};

class  AbstractFactory
{
public:
	virtual Fruit* createBanana() = 0;
	virtual Fruit* createApple() = 0;
};

class NorthFactory :public AbstractFactory
{
public:
	Fruit* createBanana()
	{
		return new NorthBanana;
	}
	Fruit* createApple()
	{
		return new NorthApple;
	}
};

class SouthFactory :public AbstractFactory
{
public:
	Fruit* createBanana()
	{
		return new SouthBanana;
	}
	Fruit* createApple()
	{
		return new SouthApple;
	}
};

int mainAbstractFactory()
{
	Fruit*				fruit = nullptr;
	AbstractFactory*	abFactory = nullptr;
	
	abFactory = new NorthFactory;
	fruit = abFactory->createApple();
	fruit->sayName();
	delete fruit;

	fruit = abFactory->createBanana();
	fruit->sayName();
	delete fruit;
	delete abFactory;

	abFactory = new SouthFactory;
	fruit = abFactory->createApple();
	fruit->sayName();
	delete fruit;

	fruit = abFactory->createBanana();
	fruit->sayName();
	delete fruit;
	delete abFactory;
	cout << "hello" << endl;
	system("pause");
	return 0;
}

File: 传统的设计模式/抽象工厂.ts
interface   Fruit{
    sayName():void; 
}

class   NorthBanana implements  Fruit{
    sayName():void{
        console.log('NorthBanana');
    }
}

class   SouthBanana implements  Fruit{
    sayName():void{
        console.log('SouthBanana');
    }
}

class   NorthApple  implements Fruit{
    sayName():void{
        console.log('NorthApple');
    }
}

class   SouthApple  implements Fruit{
    sayName():void{
        console.log('SouthApple');
    }
}

interface   AbstractFactory{
    createApple():Fruit;
    createBanana():Fruit;
}

export  class   NorthFactory    implements  AbstractFactory{
    createApple():Fruit{
        return  new NorthApple();
    }

    createBanana():Fruit{
        return  new NorthBanana();
    }
}

export  class   SouthFactory    implements  AbstractFactory{
    createApple():Fruit{
        return new SouthApple();
    }

    createBanana():Fruit{
        return new SouthBanana();
    }
}